### ISSUE TITLE
Why does not Webdriverio report an error in beforeAll with Jasmine? 

### ISSUE THREAD

**Issue by issei126** [webdriverio/webdriverio#3084]
I use with Jasmine.
When an error occur in `beforeAll`, WebdriverIO does not report the error, but exit code is 1.
I expect it to report error and test failed.
When an error occur in 'beforeEach', WebdriverIO report it and the test failed.

example code

```
describe('sample', () => {
  beforeAll(() => {
    throw new Error("ERROR")
  })

  it('sample', () => {
    browser.url('http://www.google.com')
  })
})
```

**Environment (please complete the following information):**
 - **WebdriverIO version: 4.14.0
 - **Mode:** WDIO Testrunner
 - **If WDIO Testrunner, running sync/async:"" sync
 - **Node.js version: 8.6.0
 - **NPM version:** 6.4.1
 - **Browser name and version:** Chrome 70
 - **Platform name and version: mac OS 10.13.6
 - **Additional wdio packages used (if applicable):** [e.g. @wdio/spec reporter, @wdio/selenium-standalone service]
   *    "chromedriver": "2.44.0",
   *   "wdio-chromedriver-service": "0.1.3",
   *  "wdio-jasmine-framework": "0.3.7",

**Comment by christian-bromann** [comment]
@issei126 please share the report and the report that you would expect. I don't know what you mean by report here.

**Comment by issei126** [comment]
@christian-bromann 

Sorry. The above code is not good.

Here is another alternative sample
```
describe('sample', () => {
  beforeAll(() => {
    browser.url('http://www.google.com')
    browser.click('a=amazon') // this is a sample failure operation
  })

  it('sample', () => {
    expect(browser.isExisting('a=Gmail')).toBe(true)
  })
})
```

expected report

```
F

0 passing (2.90s)
1 failing

1) samplesuite1 sample:
Failed: An element could not be located on the page using the given search parameters.
running chrome
error properties: Object({ type: 'NoSuchElement', shotTaken: true })
Error: An element could not be located on the page using the given search parameters.
    at click("a=amazon") - at element("a=amazon") - click.js:12:17
```

but report is

```
$  ./node_modules/.bin/wdio wdio.conf.js

․

1 passing 
```

**Comment by christian-bromann** [comment]
Does the Jasmine framework fail if the hook raises an Error?

**Comment by issei126** [comment]
Yes.
This is sample code

```
function add(a, b) {
    return a + b;
}


describe('sample describe', () => {
    beforeAll(() => {
        throw new Error('sample error!')
    })
    
    it('sample spec', () => {
        expect(add(1, 2)).toBe(3)
    })
})
```

output
```

~$ npm test

> ubuntu@1.0.0 test /home/ubuntu
> jasmine

DEPRECATION: Setting throwOnExpectationFailure directly on Env is deprecated, please use the oneFailurePerSpec option in `configure`
DEPRECATION: Setting randomizeTests directly is deprecated, please use the random option in `configure`
Randomized with seed 31676
Started
.


Suite error: sample describe
  Message:
    Error: sample error!
  Stack:
    Error: sample error!
        at UserContext.beforeAll (/home/ubuntu/spec/sampleSpec.js:8:15)
        at <Jasmine>
        at runCallback (timers.js:705:18)
        at tryOnImmediate (timers.js:676:5)
        at processImmediate (timers.js:658:5)

1 spec, 1 failure
Finished in 0.006 seconds
Randomized with seed 31676 (jasmine --random=true --seed=31676)
npm ERR! Test failed.  See above for more details.
```

**Comment by christian-bromann** [comment]
@issei126 feel free to provide a patch with a fix, thanks!

**Comment by xcao-riffyn** [comment]
@issei126 I tried replaced jasmine with mocha and it didn't have this issue. That tells me that it's a wdio-jasmine-framework issue rather than wdio.  Read through the code of wdio-jasmine-framework and compare it with wdio-mocha-framework now to see if I can find anything.

**Comment by christian-bromann** [comment]
I stumbled upon this issue when working on #2861 and it seems that this is a Jasmine issue. It doesn't propagate failures when an error is thrown in `beforeAll`. It works well in `beforeEach`.

**Comment by christian-bromann** [comment]
> Can you do the same fix in #2861 in wdio-mocha-framework to wdio-jasmine-framework to see if it can fix this as well?

No, that fix was related to Mocha

**Comment by xcao-riffyn** [comment]
@issei126 I dug around for some time and unfortunately didn't anything that wdio-jasmine-framework has done wrong to not to propagate the error further. Notice in your test above there was a dot underneath the Started, that indicated test passed. if you replaced beforeAll with beforeEach, that dot would have been a 'F'.  Also in wdio with wdio-jasmine-framework test I did, the test was passed but at the bottom there was a npm error 'npm ERR! Test failed.  See above for more details.'.  I did some debugging in the 
`jasmineNodeOpts: {
        //
        // Jasmine default timeout
        defaultTimeoutInterval: 10000,
        //
        // The Jasmine framework allows interception of each assertion in order to log the state of the application
        // or website depending on the result. For example, it is pretty handy to take a screenshot every time
        // an assertion fails.
        expectationResultHandler: function(passed, assertion) {
          /**
           * only take screenshot if assertion failed
           */
          console.log(`inside expectationResultHandler function block: ${passed}`);
          if(passed) {
              return;
          }

          browser.saveScreenshot('assertionError_' + assertion.error.message + '.png');
        }
    },`
with beforeAll the passed was true but with beforeEach it was false.
So somehow jasmine didn't flag error/exception in beforeAll and didn't propagate to wdio
I went through wdio-jasmine-framework's code and didn't see anything to prevent the error to progragate.

So in summary like @christian-bromann said the problem is probably at the jasmine's side

**Comment by issei126** [comment]
@xcao-riffyn 
Thank you for your investing.

I tried to use the try-catch statement in beforeAll.
It catched an error but a test passed.
I have been able to know the error, so I'd like to wait and see. 

```
describe('sample', () => {
  beforeAll(() => {
    try {
      browser.url('http://www.google.com')
      browser.click('a=amazon') // this is a sample failure operation
    }
    catch (e) {
      console.log(e)
    }
  })

  it('sample', () => {
    expect(browser.isExisting('a=Gmail')).toBe(true)
  })
})
```

output

```
$ ./node_modules/.bin/wdio wdio.conf.js

DEPRECATION: Setting randomizeTests directly is deprecated, please use the random option in `configure`
DEPRECATION: Setting specFilter directly on Env is deprecated, please use the specFilter option in `configure`
[10:11:13]      Saved screenshot: ERROR_chrome_2018-12-06T01-11-13.631Z.png
{ Error: An element could not be located on the page using the given search parameters.
    at click("a=amazon") - at element("a=amazon") - click.js:12:17
  type: 'NoSuchElement',
  message: 'An element could not be located on the page using the given search parameters.',
  shotTaken: true }
․

1 passing (4.30s)
```

**Comment by issei126** [comment]
@xcao-riffyn 
I debugged them, then I found a few interest points.

When an error occurs in beforeAll and beforeEach with wdio and jasmine, it enters the block below.
https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-sync/src/index.js#L31
So I think that jasmine may not propagate to wdio.

On the other hand,  an error occurred in beforeAll without wdio, it enters the block below. 
https://github.com/jasmine/jasmine/blob/master/lib/jasmine-core/jasmine.js#L6920
then, it enters the block below.
https://github.com/jasmine/jasmine/blob/master/lib/jasmine-core/jasmine.js#L6662
It  added as a failure there.
But an error occoured in beforeAll with wdio, it does not enter blocks above.

This may be related to this topic.

**Comment by jrobinson01** [comment]
There appears to also be a problem in `afterAll`.

**Comment by nicholasbailey** [comment]
Digging into this over the last few days, it looks like part of the problem is that at least spec-reporter is only looking for failures on hook:end and test:failed. An exception beforeEach will trigger a test:failed event in wdio-jasmine-framework, but beforeAll will not. wdio-jasmine-framework (unlike wdio-mocha-framework) never emits a hook:end event, because Jasmine's reporting interface doesn't expose the start and end of hooks.

A further complication - Jasmine catches the exception in beforeAll and doesn't propagate it immediately. So we can't even wrap beforeAll and emit a hook:end event after it is called, because the exception won't be available

I can imagine a couple of solutions here:

1. We could add a top-level notion of a suite having errors of its own, above and beyond hook and test errors. This would be a lot of work.
2. We could hook into Jasmine's Suite.onException, and use that to trigger appropriate hook:end events. The downside here is that hook:end wouldn't be emitted correctly if there were no exception. 
3. We could queue all events for a suite and emit them after the suite actually runs. We can then use the data passed in suiteDone to figure out what errors occurred in any beforeAll and afterAll hooks. 

In a vacuum, I'd do the third, but I don't know the codebase/design choices as well as some other folks here.

**Comment by nicholasbailey** [comment]
I have a possible patch for this, which I can PR for initial eyes after a little cleanup. It hooks into Jasmine.Suite's beforeAll, afterAll and onException methods and emits hook:start and hook:end events at appropriate times.  But it still leads to some reporting that's pretty counter-intuitive if you are used to Mocha.

When using Mocha, a failure in before will throw, ending suite execution. This means that you just see something like this:

![screen shot 2019-02-14 at 11 43 44 am](https://user-images.githubusercontent.com/24800147/52806131-d70bca80-304d-11e9-87fa-b71e8f94fe79.png)

The default Jasmine behavior is that failures in beforeAll _do not stop suite execution_. This means that you can get a counterintuitive scenario where you have a beforeAll failure for a suite, but individual specs within the suite are green. With the spec reporter, it looks something like this:

![screen shot 2019-02-14 at 11 44 49 am](https://user-images.githubusercontent.com/24800147/52806191-fd316a80-304d-11e9-8aa1-f75711b64db5.png)

Since this is just how Jasmine works, I don't think we want to force it to behave like Mocha. But it will almost certainly lead to cases where reporter packages spit out counter-intuitive output.

**Comment by jrobinson01** [comment]
I think I'm ok with them continuing execution. What I think is the current (my experience is mostly in v4, but I think it also exists in v5) is that errors in beforeAll, afterAll, etc are completely swallowed and don't appear in logs at all.

**Comment by mo** [comment]
The fact that beforeAll swallows all errors caused quite a bit of confusion / frustration when I upgraded from v4 to v5. I think it would be really useful to fix this issue.

**Comment by VKnightUT** [comment]
Has there been any update on this issue?

**Comment by mike-d-davydov** [comment]
> I have a possible patch for this, which I can PR for initial eyes after a little cleanup. It hooks into Jasmine.Suite's beforeAll, afterAll and onException methods and emits hook:start and hook:end events at appropriate times. But it still leads to some reporting that's pretty counter-intuitive if you are used to Mocha.
 
@nicholasbailey  Please advise if you can share that patch/MR. I would be happy to take it from there, wrap it up and contribute an MR based on your work if you have no objections. This is quite nasty bug that makes people make quite an ugly workarounds to overcome it. Thanks!

**Comment by mo** [comment]
@nicholasbailey can you post your patch even if it's not finished? if so, then maybe @mike-d-davydov can finish it

**Comment by mgrybyk** [comment]
gonna be fixed with one of PRs
